### PR TITLE
Bug Fix: Log messages from Java code base

### LINKED ISSUE #7526: Log message from Java code won't actually generate logs
Due to the way we initialize the log4j logging context via Ruby, the Java logging context results in using the default context which will only log ERROR message to stdout (and/or sterr). 

This is difficult to reproduce without injecting code for testing since Java logging is already quiet unless specific errors occur. (to reproduce just inject an artificial log message in the Event class, and watch it not show up in the logs)

### PR DESCRIPTION
**Pull request by jakelandis**
This change maps log4j2 context established in Ruby to the logging context used by Java.

Fixes #7526

### CODE REVIEW

**Comment by jakelandis** [comment]
This likely also fixes 
https://github.com/elastic/logstash/issues/4704
https://github.com/elastic/logstash/issues/4111
https://github.com/elastic/logstash/issues/7242

**Comment by ph** [comment]
@jakelandis I have merged this PR my java branch, but It doesn't seems to fix my issues.

```
[2017-06-28T16:54:21,222][INFO ][logstash.javapipeline    ] Still in ruby
Hello? Logger? # System.out.println()
[2017-06-28T16:54:21,224][INFO ][logstash.javapipeline    ] Back to ruby
```

In the Java code, I am logging at the info level with a simple statement `logger.info("starting pipeline {}", pipelineId)`

**Comment by jakelandis** [comment]
is your log statement in core or a plugin ?

Also, if you are testing from intelliJ, make sure that you also grab this minor change: https://github.com/elastic/logstash/pull/7533 to ensure your classpath is correct. _(else do a full ./gradlew vendor to pick up your changes)_

**Comment by ph** [comment]
This was in core, I didn't test it in IntelliJ and I Duda gradle assemble
to rebuild the jars I don't think it make a difference from vendor I can do
another test

**Comment by jakelandis** [comment]
Do you have a link to your class with the logger ?

**Comment by ph** [comment]
@jakelandis I've created a simple replicable case with the following changes.

I just start logstash on the command line:

```
bin/logstash -e "input { generator {} } output { null {} }" --log.level debug
```

And the only output I am receiving is this:

```
[2017-06-28T19:30:08,344][DEBUG][logstash.outputs.file    ] config LogStash::Outputs::File/@file_mode = -1
Start logging
Stop logging
[2017-06-28T19:30:08,349][DEBUG][logstash.pipeline        ] Starting pipeline {:pipeline_id=>"main"}
```

```diff
diff --git a/logstash-core/lib/logstash/pipeline.rb b/logstash-core/lib/logstash/pipeline.rb
index a5d22a7c0..4fce0a37f 100644
--- a/logstash-core/lib/logstash/pipeline.rb
+++ b/logstash-core/lib/logstash/pipeline.rb
@@ -29,6 +29,7 @@ require "securerandom"
 java_import org.logstash.common.DeadLetterQueueFactory
 java_import org.logstash.common.SourceWithMetadata
 java_import org.logstash.common.io.DeadLetterQueueWriter
+java_import org.logstash.ReplCase

 module LogStash; class BasePipeline
   include LogStash::Util::Loggable
@@ -253,6 +254,8 @@ module LogStash; class Pipeline < BasePipeline
   end

   def start
+    test = org.logstash.ReplCase.new
+    test.doSomething()
     # Since we start lets assume that the metric namespace is cleared
     # this is useful in the context of pipeline reloading
     collect_stats
```

```java
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;

public class ReplCase {
    private static final Logger logger = LogManager.getLogger(ReplCase.class);

    public ReplCase() {

    }

    public void doSomething() {
        System.out.println("Start logging");
        logger.info("Log at the info");
        logger.debug("Log at the debug");
        System.out.println("Stop logging");
    }
}

**Comment by jakelandis** [comment]
@ph Thanks for example. 

Digging into it a bit this is not working because Logger in you class is resulting in using the default context (the very bug which this PR sets to fix!). 

It is using the default context because the static initialization of your logger is occurring BEFORE we set up our own context. This class is a little bit special _(more so then the ones I was testing with)_ because it is referenced by a Ruby file (via java_include) that is executed BEFORE we setup our context.  So.... there exists two sets of log4j2 contexts (default and the one we create), and it appears that any class that is referenced BEFORE we setup our context and the logger is established at static scope (which they should be) then those loggers won't work as expected. 

I will take a look at our options here, likely try to move the logging initialization WAY up (like the very first thing) in the chain of startup events.  If not, we may need to re-think establishing our own context.

**Comment by jakelandis** [comment]
@ph - can you re-test ? (The build failure is [un-related](https://github.com/elastic/logstash/issues/7574))

I found the log4j2 magic method that allows us to re-configure the default logging context. So the change that fixes this is to no longer establish our own context, but instead just re-configure the default.  Everything seems happy now regardless of initialization ordering. _(Though non error logging from some static blocks still won't work, but that is a very edge case)_  

Reminder to self: update the commit wording when squashing since we no longer need to describe things in terms of two logging contexts.

**Comment by jakelandis** [comment]
Jenkins test this please
